File: backend-C#/Services/ChatService.cs
using IMS.Data;
using IMS.DTOs.Requests;
using IMS.DTOs.Responses;
using IMS.Models;
using Microsoft.EntityFrameworkCore;

namespace IMS.Services;

public class ChatService : IChatService
{
    private readonly ApplicationDbContext _context;

    public ChatService(ApplicationDbContext context)
    {
        _context = context;
    }

    public async Task<ChatListResponse> GetChatListAsync(long currentUserId)
    {
        // Get all other active users as contacts
        var users = await _context.Users
            .Where(u => u.Id != currentUserId && u.IsActive)
            .ToListAsync();

        var contacts = new List<ChatContactResponse>();

        foreach (var user in users)
        {
            var contact = new ChatContactResponse
            {
                Id = user.Id,
                UserId = user.Id,
                Name = user.FullName ?? user.Username,
                Email = user.Email,
                Avatar = user.Avatar
            };

            // Get last message
            var lastMsg = await _context.ChatMessages
                .Where(m => (m.SenderId == currentUserId && m.ReceiverId == user.Id) ||
                           (m.SenderId == user.Id && m.ReceiverId == currentUserId))
                .OrderByDescending(m => m.CreatedAt)
                .FirstOrDefaultAsync();

            if (lastMsg != null)
            {
                contact.LastMessage = lastMsg.Content;
                contact.LastMessageTime = lastMsg.CreatedAt.ToString("yyyy-MM-dd HH:mm:ss");
            }

            // Get unread count
            contact.UnreadCount = await _context.ChatMessages
                .CountAsync(m => m.SenderId == user.Id && m.ReceiverId == currentUserId && !m.IsRead);

            contacts.Add(contact);
        }

        return new ChatListResponse
        {
            Contacts = contacts,
            Total = contacts.Count
        };
    }

    public async Task<ChatHistoryResponse> GetChatHistoryAsync(long currentUserId, long otherUserId, ChatHistoryRequest request)
    {
        var page = request.Page < 1 ? 1 : request.Page;
        var limit = request.Limit < 1 || request.Limit > 100 ? 50 : request.Limit;
        var offset = (page - 1) * limit;

        // Count total
        var total = await _context.ChatMessages
            .CountAsync(m => (m.SenderId == currentUserId && m.ReceiverId == otherUserId) ||
                            (m.SenderId == otherUserId && m.ReceiverId == currentUserId));

        // Get messages
        var messages = await _context.ChatMessages
            .Where(m => (m.SenderId == currentUserId && m.ReceiverId == otherUserId) ||
                       (m.SenderId == otherUserId && m.ReceiverId == currentUserId))
            .OrderByDescending(m => m.CreatedAt)
            .Skip(offset)
            .Take(limit)
            .ToListAsync();

        var response = new List<ChatMessageResponse>();
        foreach (var msg in messages.OrderBy(m => m.CreatedAt))
        {
            var sender = await _context.Users
                .Select(u => new { u.Id, u.FullName, u.Username })
                .FirstOrDefaultAsync(u => u.Id == msg.SenderId);

            var receiver = await _context.Users
                .Select(u => new { u.Id, u.FullName, u.Username })
                .FirstOrDefaultAsync(u => u.Id == msg.ReceiverId);

            response.Add(new ChatMessageResponse
            {
                Id = msg.Id,
                Content = msg.Content,
                SenderId = msg.SenderId,
                SenderName = sender?.FullName ?? sender?.Username ?? "",
                ReceiverId = msg.ReceiverId,
                ReceiverName = receiver?.FullName ?? receiver?.Username ?? "",
                IsRead = msg.IsRead,
                CreatedAt = msg.CreatedAt.ToString("yyyy-MM-dd HH:mm:ss")
            });
        }

        return new ChatHistoryResponse
        {
            Messages = response,
            Total = total,
            Page = page,
            Limit = limit
        };
    }

    public async Task<ChatMessageResponse> SendMessageAsync(long currentUserId, SendMessageRequest request)
    {
        // Check receiver exists
        var receiver = await _context.Users.FindAsync(request.ReceiverId);
        if (receiver == null)
        {
            throw new InvalidOperationException("接收者不存在");
        }

        var message = new ChatMessage
        {
            SenderId = currentUserId,
            ReceiverId = request.ReceiverId,
            Content = request.Content,
            IsRead = false,
            CreatedAt = DateTime.UtcNow,
            UpdatedAt = DateTime.UtcNow
        };

        _context.ChatMessages.Add(message);
        await _context.SaveChangesAsync();

        var sender = await _context.Users
            .Select(u => new { u.Id, u.FullName, u.Username })
            .FirstOrDefaultAsync(u => u.Id == currentUserId);

        return new ChatMessageResponse
        {
            Id = message.Id,
            Content = message.Content,
            SenderId = message.SenderId,
            SenderName = sender?.FullName ?? sender?.Username ?? "",
            ReceiverId = message.ReceiverId,
            ReceiverName = receiver.FullName ?? receiver.Username ?? "",
            IsRead = message.IsRead,
            CreatedAt = message.CreatedAt.ToString("yyyy-MM-dd HH:mm:ss")
        };
    }

    public async Task<long> MarkAsReadAsync(long currentUserId, long otherUserId)
    {
        var messages = await _context.ChatMessages
            .Where(m => m.SenderId == otherUserId && m.ReceiverId == currentUserId && !m.IsRead)
            .ToListAsync();

        foreach (var msg in messages)
        {
            msg.IsRead = true;
        }

        await _context.SaveChangesAsync();
        return messages.Count;
    }

    public async Task<long> GetUnreadCountAsync(long currentUserId)
    {
        return await _context.ChatMessages
            .CountAsync(m => m.ReceiverId == currentUserId && !m.IsRead);
    }

    public async Task<ChatHistoryResponse> GetAllMessagesAsync(long currentUserId, int page, int limit, string? search)
    {
        page = page < 1 ? 1 : page;
        limit = limit < 1 || limit > 100 ? 50 : limit;
        var offset = (page - 1) * limit;

        var query = _context.ChatMessages.AsQueryable();

        if (!string.IsNullOrEmpty(search))
        {
            query = query.Where(m => EF.Functions.Like(m.Content, $"%{search}%"));
        }

        var total = await query.CountAsync();

        var messages = await query
            .OrderByDescending(m => m.CreatedAt)
            .Skip(offset)
            .Take(limit)
            .ToListAsync();

        var response = new List<ChatMessageResponse>();
        foreach (var msg in messages)
        {
            var sender = await _context.Users
                .Select(u => new { u.Id, u.FullName, u.Username })
                .FirstOrDefaultAsync(u => u.Id == msg.SenderId);

            var receiver = await _context.Users
                .Select(u => new { u.Id, u.FullName, u.Username })
                .FirstOrDefaultAsync(u => u.Id == msg.ReceiverId);

            response.Add(new ChatMessageResponse
            {
                Id = msg.Id,
                Content = msg.Content,
                SenderId = msg.SenderId,
                SenderName = sender?.FullName ?? sender?.Username ?? "",
                ReceiverId = msg.ReceiverId,
                ReceiverName = receiver?.FullName ?? receiver?.Username ?? "",
                IsRead = msg.IsRead,
                CreatedAt = msg.CreatedAt.ToString("yyyy-MM-dd HH:mm:ss")
            });
        }

        return new ChatHistoryResponse
        {
            Messages = response,
            Total = total,
            Page = page,
            Limit = limit
        };
    }

    public async Task<ChatListResponse> GetAllConversationsAsync(long currentUserId, int page, int limit, string? search)
    {
        page = page < 1 ? 1 : page;
        limit = limit < 1 || limit > 100 ? 50 : limit;
        var offset = (page - 1) * limit;

        var query = _context.Users.Where(u => u.IsActive).AsQueryable();

        if (!string.IsNullOrEmpty(search))
        {
            query = query.Where(u => 
                EF.Functions.Like(u.Username, $"%{search}%") ||
                (u.FullName != null && EF.Functions.Like(u.FullName, $"%{search}%")) ||
                EF.Functions.Like(u.Email, $"%{search}%"));
        }

        var total = await query.CountAsync();

        var users = await query
            .Skip(offset)
            .Take(limit)
            .ToListAsync();

        var conversations = new List<ChatContactResponse>();
        foreach (var user in users)
        {
            var name = user.FullName ?? user.Username;

            // Get last message
            var lastMsg = await _context.ChatMessages
                .Where(m => m.SenderId == user.Id || m.ReceiverId == user.Id)
                .OrderByDescending(m => m.CreatedAt)
                .FirstOrDefaultAsync();

            var unreadCount = await _context.ChatMessages
                .CountAsync(m => m.ReceiverId == user.Id && !m.IsRead);

            conversations.Add(new ChatContactResponse
            {
                Id = user.Id,
                UserId = user.Id,
                Name = name,
                Email = user.Email,
                Avatar = user.Avatar,
                LastMessage = lastMsg?.Content,
                LastMessageTime = lastMsg?.CreatedAt.ToString("yyyy-MM-dd HH:mm:ss"),
                UnreadCount = unreadCount
            });
        }

        return new ChatListResponse
        {
            Contacts = conversations,
            Total = total
        };
    }
}

File: backend-JS/src/services/chatService.js
const { ChatMessage, User, Sequelize, sequelize } = require('../models');

const getChatList = async (currentUserId) => {
  const users = await User.findAll({
    where: {
      id: { [Sequelize.Op.ne]: currentUserId },
      isActive: true
    },
    attributes: ['id', 'username', 'email', 'fullName', 'avatar']
  });

  const contacts = await Promise.all(users.map(async (user) => {
    const lastMessage = await ChatMessage.findOne({
      where: {
        [Sequelize.Op.or]: [
          { senderId: currentUserId, receiverId: user.id },
          { senderId: user.id, receiverId: currentUserId }
        ]
      },
      order: [['createdAt', 'DESC']]
    });

    const unreadCount = await ChatMessage.count({
      where: {
        senderId: user.id,
        receiverId: currentUserId,
        isRead: false
      }
    });

    return {
      id: user.id,
      userId: user.id,
      name: user.fullName || user.username,
      email: user.email,
      avatar: user.avatar,
      lastMessage: lastMessage ? lastMessage.content : '',
      lastMessageTime: lastMessage ? lastMessage.createdAt.toISOString().replace('T', ' ').substring(0, 19) : '',
      unreadCount
    };
  }));

  return contacts;
};

const getChatHistory = async (currentUserId, otherUserId, page = 1, limit = 50) => {
  const offset = (page - 1) * limit;

  const { count, rows } = await ChatMessage.findAndCountAll({
    where: {
      [Sequelize.Op.or]: [
        { senderId: currentUserId, receiverId: otherUserId },
        { senderId: otherUserId, receiverId: currentUserId }
      ]
    },
    order: [['createdAt', 'DESC']],
    limit: parseInt(limit),
    offset: parseInt(offset),
    include: [
      { model: User, as: 'Sender', attributes: ['id', 'username', 'fullName'] },
      { model: User, as: 'Receiver', attributes: ['id', 'username', 'fullName'] }
    ]
  });

  const messages = rows.reverse().map(msg => ({
    id: msg.id,
    content: msg.content,
    senderId: msg.senderId,
    senderName: msg.Sender ? (msg.Sender.fullName || msg.Sender.username) : '',
    receiverId: msg.receiverId,
    receiverName: msg.Receiver ? (msg.Receiver.fullName || msg.Receiver.username) : '',
    isRead: msg.isRead,
    createdAt: msg.createdAt.toISOString().replace('T', ' ').substring(0, 19)
  }));

  return {
    messages,
    total: count,
    page: parseInt(page),
    limit: parseInt(limit)
  };
};

const sendMessage = async (currentUserId, receiverId, content) => {
  const message = await ChatMessage.create({
    senderId: currentUserId,
    receiverId: receiverId,
    content,
    isRead: false
  });

  const sender = await User.findByPk(currentUserId);
  const receiver = await User.findByPk(receiverId);

  return {
    id: message.id,
    content: message.content,
    senderId: message.senderId,
    senderName: sender ? (sender.fullName || sender.username) : '',
    receiverId: message.receiverId,
    receiverName: receiver ? (receiver.fullName || receiver.username) : '',
    isRead: message.isRead,
    createdAt: message.createdAt.toISOString().replace('T', ' ').substring(0, 19)
  };
};

const markAsRead = async (currentUserId, otherUserId) => {
  await ChatMessage.update(
    { isRead: true },
    {
      where: {
        senderId: otherUserId,
        receiverId: currentUserId,
        isRead: false
      }
    }
  );
  return { success: true };
};

const getUnreadCount = async (currentUserId) => {
  const count = await ChatMessage.count({
    where: {
      receiverId: currentUserId,
      isRead: false
    }
  });
  return { count };
};

const getAllMessages = async (page = 1, limit = 20) => {
  const offset = (page - 1) * limit;

  const { count, rows } = await ChatMessage.findAndCountAll({
    order: [['createdAt', 'DESC']],
    limit: parseInt(limit),
    offset: parseInt(offset),
    include: [
      { model: User, as: 'Sender', attributes: ['id', 'username', 'fullName'] },
      { model: User, as: 'Receiver', attributes: ['id', 'username', 'fullName'] }
    ]
  });

  const messages = rows.map(msg => ({
    id: msg.id,
    content: msg.content,
    senderId: msg.senderId,
    senderName: msg.Sender ? (msg.Sender.fullName || msg.Sender.username) : '',
    receiverId: msg.receiverId,
    receiverName: msg.Receiver ? (msg.Receiver.fullName || msg.Receiver.username) : '',
    isRead: msg.isRead,
    createdAt: msg.createdAt.toISOString().replace('T', ' ').substring(0, 19)
  }));

  return {
    messages,
    total: count,
    page: parseInt(page),
    limit: parseInt(limit)
  };
};

const getAllConversations = async (page = 1, limit = 50) => {
  try {
    // 验证参数
    const pageNum = parseInt(page) || 1;
    const limitNum = parseInt(limit) || 50;
    
    // 获取所有消息，按创建时间倒序排序
    const messages = await ChatMessage.findAll({
      order: [['createdAt', 'DESC']],
      include: [
        { model: User, as: 'Sender', attributes: ['id', 'username', 'fullName'] },
        { model: User, as: 'Receiver', attributes: ['id', 'username', 'fullName'] }
      ]
    });
    
    // 构建唯一对话映射，避免重复
    const conversationMap = new Map();
    
    for (const message of messages) {
      // 确保消息对象有效
      if (!message || !message.senderId || !message.receiverId) {
        continue;
      }
      
      // 创建唯一对话键，确保顺序一致（小ID在前，大ID在后）
      const user1Id = Math.min(message.senderId, message.receiverId);
      const user2Id = Math.max(message.senderId, message.receiverId);
      const conversationKey = `${user1Id}-${user2Id}`;
      
      // 如果对话不存在，创建新对话
      if (!conversationMap.has(conversationKey)) {
        const sender = message.Sender;
        const receiver = message.Receiver;
        
        const user1 = user1Id === message.senderId ? sender : receiver;
        const user2 = user2Id === message.senderId ? sender : receiver;
        
        conversationMap.set(conversationKey, {
          key: conversationKey,
          user1Id,
          user2Id,
          user1Name: user1 ? (user1.fullName || user1.username || '未知用户') : '未知用户',
          user2Name: user2 ? (user2.fullName || user2.username || '未知用户') : '未知用户',
          messageCount: 0,
          lastMessage: message.content || '',
          lastMessageTime: message.createdAt && !isNaN(new Date(message.createdAt)) ? new Date(message.createdAt).toISOString().replace('T', ' ').substring(0, 19) : ''
        });
      }
      
      // 增加消息计数
      const conversation = conversationMap.get(conversationKey);
      if (conversation) {
        conversation.messageCount++;
      }
    }
    
    // 将映射转换为数组，并按最后消息时间倒序排序
    const conversations = Array.from(conversationMap.values())
      .sort((a, b) => {
        const dateA = a.lastMessageTime ? new Date(a.lastMessageTime) : new Date(0);
        const dateB = b.lastMessageTime ? new Date(b.lastMessageTime) : new Date(0);
        return dateB - dateA;
      });
    
    // 实现分页
    const startIndex = (pageNum - 1) * limitNum;
    const endIndex = startIndex + limitNum;
    const paginatedConversations = conversations.slice(startIndex, endIndex);
    
    return paginatedConversations;
  } catch (error) {
    console.error('Error in getAllConversations:', error);
    throw error;
  }
};

module.exports = {
  getChatList,
  getChatHistory,
  sendMessage,
  markAsRead,
  getUnreadCount,
  getAllMessages,
  getAllConversations
};
